presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java

```diff
@@ -2042,21 +2049,32 @@ public String getTableName()
 
         public void addPartition(Partition partition)
         {
+            checkArgument(getPrestoQueryId(partition).isPresent());
             partitions.add(partition);
         }
 
         public void execute()
         {
             List<List<Partition>> batchedPartitions = Lists.partition(partitions, batchSize);
             for (List<Partition> batch : batchedPartitions) {
-                metastore.addPartitions(schemaName, tableName, batch);
-                // One of the possible reasons that addPartitions fail is that the partition may already exist.
-                // Therefore, `batch` cannot be added to the list of created partitions before addPartitions call returns.
-                // However, adding it after could result in transaction not fully rolled back.
-                // But there isn't any guarantee that metastore will respond to our dropPartition calls anyways.
-                // On the other hand, `rollback` cannot guarantee that it only deletes partition we added.
-                for (Partition partition : batch) {
-                    createdPartitionValues.add(partition.getValues());
+                try {
+                    metastore.addPartitions(schemaName, tableName, batch);
+                }
+                finally {
+                    // Add partition to the created list conservatively
+                    // Some metastore implementations are known to violate all or none guarantee for add_partitions call.
```

Comment:
I think it would be clearer to says `Some metastore implementations are known to violate the "all or none" guarantee for add_partitions call.`